Updating the plugin docs to the V2 Android Plugin APIs

sfshaza2 opened the pull request:
Please review! I made a pass through all of the pages in the Packages & Plugins directory. I updated what I could, based on my reading of Chris' blog post and other docs. But please let me know if I've written any fiction!

Also, I have a few PENDING comments and I'd appreciate if your can collectively help me fill those sections in.

Look at all the pages in the Packages & Plugins directory. It's staged here:
https://sz-flutter.firebaseapp.com/docs/development/packages-and-plugins/using-packages

Thanks!

ditman commented on src/docs/development/platform-integration/c-interop.md:
```diff
@@ -350,7 +350,7 @@ in binary form, use the following instructions:
 ### Web
 
-Plugins are not yet supported for web apps.
+PENDING
```
AFAIK ffi is not supported in Web.

sfshaza2 replied:
Thx

ditman commented on src/docs/development/platform-integration/platform-channels.md:
```diff
@@ -4,13 +4,15 @@ short-title: Platform-specific code
 description: Learn how to write custom platform-specific code in your app.
 ---
 
+[PENDING: Is this page up-to-date?]
```
This is still interesting, particularly the part about data serialization and supported types. I think this is still current and valid for all "native" plugins. Only "web in a federated architecture" doesn't use the platform channels.

sfshaza2 replied:
Thanks!

ditman commented on src/docs/development/platform-integration/platform-channels.md:
```diff
@@ -632,47 +630,37 @@ and the app displays 'battery info unavailable'.
+To share your platform-specific code with other developers
+in the Flutter ecosystem, see [publishing packages][].
 
 ## Custom channels and codecs
```
There's an example of a custom codec in the cloud_firestore plugin, that is able to serialize/deserialize many more types than the default ones, here:

https://github.com/FirebaseExtended/flutterfire/blob/master/packages/cloud_firestore/cloud_firestore_platform_interface/lib/src/method_channel/utils/firestore_message_codec.dart

sfshaza2 replied:
Excellent!

sfshaza2 commented on src/docs/cookbook/forms/focus.md:
```diff
@@ -147,7 +147,7 @@ FloatingActionButton(
   it fails due to [Issue 52221][] when run against the 
   stable (1.12) or beta (1.14) versions of the Flutter SDK.
-  Until this [fix has landed][] (hash [flutter/flutter@bf551a3])[]
+  Until this [fix has landed][] (hash [flutter/flutter@bf551a3][])
```
:D

sfshaza2 commented on src/docs/development/packages-and-plugins/developing-packages.md:
```diff
@@ -4,218 +4,333 @@ short-title: Developing
+**`test/hello_test.dart`**
+: The [unit tests][] for the package.
+
+[PENDING: Is this still true for regular Dart packages?]
```
D'OH. Why didn't I think of that! Thanks. My next commit addresses this.

sfshaza2 commented on src/docs/development/packages-and-plugins/developing-packages.md:
```diff
@@ -4,218 +4,333 @@ short-title: Developing
+or in several files in the `lib` directory.
 
-To test the package, add [unit tests]
+To test the package, add [unit tests][]
```
It's our convention. The empty brackets aren't technically necessary, but we were told it's a best practice for some reason I can't remember. So we add them in when we see them.

sfshaza2 commented on src/docs/development/packages-and-plugins/developing-packages.md:
```diff
@@ -4,218 +4,333 @@ short-title: Developing
+Android (Kotlin or Java code), iOS (Swift or Objective-C),
+web (Dart), macos (Dart), or any subset thereof.
+The API is connected to the platform-specific
+implementation(s) using [platform channels][].
```
Nice!

sfshaza2 commented on src/docs/development/packages-and-plugins/developing-packages.md:
```diff
@@ -4,218 +4,333 @@ short-title: Developing
+**platform package(s)**
+: One or more packages that contain the platform-specific
+  implementation code. The app-facing package calls into
+  these packages&mdash;they aren't included into an app.
```
thx

sfshaza2 commented on src/docs/development/packages-and-plugins/developing-packages.md:
```diff
@@ -225,45 +340,66 @@ It is recommended practice to add the following documentation to all packages:
 
-1. Change directory to the location of your package:
+<ol markdown="1">
+<li markdown="1">Change directory to the location of your package:
```
It basically breaks down when you start adding code blocks, images, etc.

sfshaza2 commented on src/docs/development/packages-and-plugins/developing-packages.md:
```diff
@@ -4,218 +4,370 @@ short-title: Developing
 
 ## Developing Dart packages {#dart}
 
+The following instructions explain how to write a Dart package
```
Very good point. Thanks for catching that.

sfshaza2 commented on src/docs/development/packages-and-plugins/developing-packages.md:
```diff
@@ -4,218 +4,370 @@ short-title: Developing
 
+{{site.note.alert}}
+  If you write plugins for Flutter, you should know that
+  the Android plugin API was upgraded to 2.0 in Flutter 1.12
```
Done

sfshaza2 commented on src/docs/development/packages-and-plugins/developing-packages.md:
```diff
@@ -4,218 +4,333 @@ short-title: Developing
+A federated plugin requires the following packages:
+
+**app-facing package**
+: A mandatory package that you include in your Dart code
```
Done

sfshaza2 commented on src/docs/development/packages-and-plugins/developing-packages.md:
```diff
@@ -4,218 +4,333 @@ short-title: Developing
+  these packages&mdash;they aren't included into an app.
+
+**platform interface package**
+: The mandatory package that glues the app-facing packing
```
Done

sfshaza2 commented on src/docs/development/packages-and-plugins/developing-packages.md:
```diff
@@ -4,218 +4,333 @@ short-title: Developing
+  sdk: ">=2.1.0 <3.0.0"
+  # Flutter versions prior to 1.10 did not support the
+  # flutter.plugin.platforms map.
+  flutter: ">=1.10.0 <2.0.0"
```
ok

sfshaza2 commented on src/docs/development/packages-and-plugins/developing-packages.md:
```diff
@@ -4,218 +4,333 @@ short-title: Developing
-        pluginClass: HelloPlugin
-      ios:
-        pluginClass: HelloPlugin
+PENDING
```
Done

sfshaza2 commented:
@ditman, @johnpryan, I believe I've addressed all of your feedback. Thanks so much! Any reason that I can't merge this now? (All of the PENDING comments are removed or commented out.)

sfshaza2 commented:
Thanks!
